[quant][fx] Add lowering support for ops in GeneralTensorShapeOpQuantizeHandler

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #72490
* #72489
* #72488
* #72487
* __->__ #72387

Summary:
Also make GeneralTensorShapeOpQuantizeHandler produce reference patterns by default

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34025005](https://our.internmc.facebook.com/intern/diff/D34025005)

jerryzh168 committed:
[quant][fx] Add lowering support for ops in GeneralTensorShapeOpQuantizeHandler

Summary:
Also make GeneralTensorShapeOpQuantizeHandler produce reference patterns by default

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34025005).

jerryzh168 commented on test/quantization/fx/test_quantize_fx.py:
```diff
@@ -5458,7 +5458,7 @@ def forward(self, x):
         m = M(w, b).eval()
         qconfig_dict = {
-            # this has no effect on reshape since it's a CopyNode
+            # reshape will be quantized to fp16 as requested by this qconfig
```
@vkuzo I think we can rely on default_reuse_input_config now to configure reshape op to inherit the qconfig from input, by default it can use what is configured for that op, let me know what do you think

vkuzo replied:
sorry, not sure what the question is, could you clarify?

jerryzh168 replied:
nvm, I thought I changed the behavior in this PR, but turns out I didn't...
this test case is the one that we have a global fp16 config for reshape, and checks if it has any effect on reshape when it is surrounded by int8 ops. current behavior is that we would respect the qconfig.

vkuzo commented on torch/ao/quantization/fx/_lower_to_native_backend.py:
```diff
@@ -100,14 +101,24 @@ def special_pattern_replacement(model: QuantizedGraphModule) -> QuantizedGraphMo
     for n in model.graph.nodes:
         q_node = n
-        if q_node.target == torch.quantize_per_tensor:
+        if q_node.target == torch.quantize_per_tensor or \
```
optional: maybe we can do 

```
if not statement:
    continue
...logic...
```
to save some indent, here and also in L121?

jerryzh168 replied:
sg

andrewor14 commented on torch/ao/quantization/fx/_lower_to_native_backend.py:
```diff
@@ -100,14 +101,24 @@ def special_pattern_replacement(model: QuantizedGraphModule) -> QuantizedGraphMo
+                elif isinstance(dq_node, (tuple, list)):
+                    is_dequantize = all(x.target == 'dequantize' for x in dq_node)
+                else:
+                    is_dequantize = False
```
Do we ever hit this case since `dq_node` must be Node or tuple or list (assert in L114)?

jerryzh168 replied:
yeah.. we won't hit this, I can change to initialize is_dquantize to False I think, in case we missed it in the branches in the future

andrewor14 commented on torch/ao/quantization/fx/_lower_to_native_backend.py:
```diff
@@ -118,16 +129,23 @@ def special_pattern_replacement(model: QuantizedGraphModule) -> QuantizedGraphMo
+                        for dn in dq_node:
+                            dn_input = dn.args[0]
+                            dn.replace_all_uses_with(dn_input)
+                            model.graph.erase_node(dn)
```
Maybe we can do something like the following to reduce the duplicate code?
```
dq_nodes = []
if isinstance(dq_node, Node):
    dq_nodes = [dq_node.args[0]]
elif isinstance(dq_node, (tuple, list)):
    dq_nodes = dq_node.args[0]

for dn in dq_nodes:
    ... # replace the node from the graph
```

jerryzh168 replied:
makes sense, will do

andrewor14 commented on torch/ao/quantization/fx/quantization_patterns.py:
```diff
@@ -1733,31 +1733,28 @@ def convert(self,
                 # TODO: remove special case for operator.getitem
                 # make sure the input is quantized to act_dtype if it's not operator.getitem
-                if node.target != operator.getitem:
-                    load_arg(quantized={0: act_dtype})(node.args)
```
should we also remove the TODO comment above?

jerryzh168 replied:
yeah sure

jerryzh168 committed:
Update on "[quant][fx] Add lowering support for ops in GeneralTensorShapeOpQuantizeHandler"


Summary:
Also make GeneralTensorShapeOpQuantizeHandler produce reference patterns by default

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34025005](https://our.internmc.facebook.com/intern/diff/D34025005)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34025005).

terrychenism reviewed (commented):
LGTM

jerryzh168 committed:
Update on "[quant][fx] Add lowering support for ops in GeneralTensorShapeOpQuantizeHandler"


Summary:
Also make GeneralTensorShapeOpQuantizeHandler produce reference patterns by default

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34025005](https://our.internmc.facebook.com/intern/diff/D34025005)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34025005).
